pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,134 @@
+    >>> def circuit():
+    >>>     qml.Hadamard(wires=[0])
+    >>>     qml.Toffoli(wires=[0,1,2])
+    >>>     return qml.expval(qml.Z(0))
+    >>>
+    >>> print(qml.draw(circuit)())
+    0: ──RZ(1.57)──RX(1.57)──RZ(1.57)─╭●─┤  <Z>
+    1: ───────────────────────────────├●─┤
+    2: ───────────────────────────────╰X─┤
+
+    >>> @partial(decompose, gate_set=lambda op: len(op.wires) <= 2)
+    >>> @qml.qnode(dev)
+    >>> def circuit():
+    >>>     qml.Hadamard(wires=[0])
+    >>>     qml.Toffoli(wires=[0,1,2])
+    >>>     return qml.expval(qml.Z(0))
+    >>>
+    >>> print(qml.draw(circuit)())
+    0: ──H────────╭●───────────╭●────╭●──T──╭●─┤  <Z>
+    1: ────╭●─────│─────╭●─────│───T─╰X──T†─╰X─┤
+    2: ──H─╰X──T†─╰X──T─╰X──T†─╰X──T──H────────┤
+
+    >>> tape = qml.tape.QuantumScript([qml.IsingXX(1.2, wires=(0,1))], [qml.expval(qml.Z(0))])
+    >>> batch, fn = qml.transforms.decompose(tape, gate_set={"CNOT", "RX", "RZ"})
+    >>> batch[0].circuit
+    [CNOT(wires=[0, 1]), RX(1.2, wires=[0]), CNOT(wires=[0, 1]), expval(Z(0))]
+
+    """
+
+    if gate_set is None:
+        gate_set = set(qml.ops.__all__)
+
+    if isinstance(gate_set, (str, type)):
+        gate_set = set([gate_set])
+
+    if isinstance(gate_set, (list, tuple)):
+        gate_set = set(gate_set)
+
+    if isinstance(gate_set, set):
+        gate_types = tuple(gate for gate in gate_set if isinstance(gate, type))
+        gate_names = set(gate for gate in gate_set if isinstance(gate, str))
+        gate_set = lambda op: (op.name in gate_names) or isinstance(op, gate_types)
+
+    def decomposer(op):
+        return op.decomposition()
+
+    def stopping_condition(op):
+        if not op.has_decomposition:
+            if not gate_set(op):
+                warnings.warn(
+                    f"Operator {op.name} has no supported decomposition and was not found in the set of allowed decomposition gates."
+                    f"To remove this warning, add the operator name ({op.name}) or type ({type(op)}) to the allowed set of gates.",
+                    UserWarning,
+                )
+            return True
+        return gate_set(op)
+
+    if all(stopping_condition(op) for op in tape.operations):
+        return (tape,), null_postprocessing
+
+    try:
+        new_ops = [
+            final_op
+            for op in tape.operations
+            for final_op in qml.devices.preprocess._operator_decomposition_gen(
+                op,
+                stopping_condition,
+                decomposer,
+                max_expansion=max_expansion,
+            )
+        ]
+    except RecursionError as e:
+        raise RecursionError(
+            "Reached recursion limit trying to decompose operations. "
+            "Operator decomposition may have entered an infinite loop."
+            "Setting ``max_expansion`` will terminate the decomposition after a set number."
+        ) from e
+
+    tape = QuantumScript(new_ops, tape.measurements, shots=tape.shots)
```

Comment:
Can potentially use Lillian's recent improvement and do `tape.copy(operations=new_ops` now  :)

Reply:
Nice!! 💯